### PR TITLE
[RDY] Amend discussion and recommendations with missing sections.

### PR DESCRIPTION
**Pull request by gfokkema**

### CODE REVIEW

File: final/recommendations/eclipse.tex [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 The delivered Eclipse frontend is just that: a frontend to the REPL backend.
 Integration with Eclipse and Spoofax Eclipse are not yet present. This
-subsection provides several recommendations to better integrate this frontend
+section provides several recommendations to better integrate this frontend
```
Comment (Hjdskes) [inline final/recommendations/eclipse.tex:7]: Whoops, that shouldn't have been here :P Oh well.

File: final/discussion/wrapping.tex [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
+
+However, as the project developed and DynSem was reimplemented significantly,
+the implemented solution has changed quite drastically. Since the
+parser employed by Spoofax uses integrated error recovery and can therefore
+automatically complement incomplete programs, ``wrapping'' turned out to be
+unnecessary. Instead, the ``ShellFacet'' introduced in \cref{sec:esv-extensions}
+allows the language designer to configure several symbols which are
+considered valid start symbols in the context of a REPL.
+
+Together with the additional shell specific rules used to obtain and
```
Comment (Balletie) [inline final/discussion/wrapping.tex]: REPL-specific

File: final/discussion/wrapping.tex [hunk 3]
```diff
@@ -1,6 +1,25 @@
 \section{Wrapping}
 \label{sec:wrapping}
 
+During the research phase of the project the investigated solutions were
+primarily focused on ``wrapping'' a partial program such that it becomes a
+valid program as a whole, as the OpenJDK team has done for
+JShell\cref{ssec:a-execution-model-repl}. Afterwards a solution using templates
```
Comment (Hjdskes) [inline final/discussion/wrapping.tex]: > JShell (see \cref{ssec:a-execution-model-repl}).

Comment (Hjdskes) [inline final/discussion/wrapping.tex]: Remove `afterwards`.

File: final/discussion/wrapping.tex [hunk 4]
```diff
@@ -1,6 +1,25 @@
...
+for various language constructs was proposed to keep this approach generic
+across all languages supported by Spoofax.
+
+However, as the project developed and DynSem was reimplemented significantly,
+the implemented solution has changed quite drastically. Since the
```
Comment (Hjdskes) [inline final/discussion/wrapping.tex]: We didn't have a solution implemented for wrapping?

File: final/discussion/wrapping.tex [hunk 5]
```diff
@@ -2,25 +2,38 @@ \section{Wrapping}
...
-primarily focused on ``wrapping'' a partial program such that it becomes a
+primarily focused on ``wrapping'' an incomplete program such that it becomes a
 valid program as a whole, as the OpenJDK team has done for JShell (see
 \cref{ssec:a-execution-model-repl}). A solution using templates for various
 language constructs was proposed to keep this approach generic across all
-languages supported by Spoofax.
+languages supported by Spoofax. \Cref{fig:wrapping} shows a diagram which
+describes this idea. The diagram shows both the wrapping of environment
+definitions and of the incomplete program AST. The environment definitions, in
+this case, refer to the language-specific constructs such as class, function and
```
Comment (Hjdskes) [inline final/discussion/wrapping.tex]: `classes, functions` 

Plural :)

File: final/discussion/wrapping.tex [hunk 6]
```diff
@@ -2,25 +2,38 @@ \section{Wrapping}
...
+  \label{fig:wrapping}
+\end{figure}
 
-Since individual DynSem rules can be retrieved from the language specification,
-these rules can then already be used to interpret the results of partial ASTs.
-The custom start symbols together with the REPL-specific rules used to obtain
-and pass execution contexts to DynSem then already results in a fully functional
-REPL, without the need to wrap partial programs inside predefined templates.
+However, as the project developed and the entry point to DynSem was
+reimplemented and refactored significantly (see \cref{sec:dynsem-refactor}),
```
Comment (Hjdskes) [inline final/discussion/wrapping.tex]: `reimplemented and refactored` is sort of the same, I'd use only one (refactored)

File: final/discussion/wrapping.tex [hunk 7]
```diff
@@ -2,25 +2,38 @@ \section{Wrapping}
...
+wrapping turned out to be an unnecessary and inadequate solution for two
+reasons. Firstly, the ability to invoke a rule corresponding to any AST, both
+complete and incomplete, automatically allows for evaluating incomplete programs
+without wrapping them first to form a complete program. Secondly, due to the
+ability of passing the semantic components as arguments to a reduction rule, the
```
Comment (Hjdskes) [inline final/discussion/wrapping.tex]: `ability to pass`

File: final/discussion/dynsem-refactor.tex [hunk 8]
```diff
@@ -0,0 +1,28 @@
+\section{Refactoring DynSem's Entry Point}
+\label{sec:dynsem-refactor}
+During the project, the implementation of DynSem had changed significantly:
```
Comment (Hjdskes) [inline final/discussion/dynsem-refactor.tex]: Remove `had`

File: final/recommendations/backend.tex [hunk 9]
```diff
@@ -30,6 +30,22 @@ \subsection{More extensive history functionality}
 
 \subsection{Analysis in context}
 
+In \cref{sec:dynsem-refactor} it was explained how DynSem was refactored to
```
Comment (Hjdskes) [inline final/recommendations/backend.tex]: > \Cref{sec:dynsem-refactor} explained ...

File: final/recommendations/backend.tex [hunk 10]
```diff
@@ -30,6 +30,22 @@ \subsection{More extensive history functionality}
...
+allow passing in an existing context in which REPL commands should be
+evaluated. Besides evaluating a complete or partial program, the REPL also
+offers the user a command to analyze a program without evaluating it.
```
Comment (Hjdskes) [inline final/recommendations/backend.tex]: Analysis in context should also work when evaluating an expression, shouldn't it?

File: final/discussion/dynsem-data-repr.tex [hunk 11]
```diff
@@ -0,0 +1,22 @@
+\section{DynSem Data Types for Rule Results}
+\label{sec:dynsem-data-repr}
+The data type of the results of the invocation of a reduction rule, are
+currently just \texttt{Object}. The actual type of the semantic components is
+unknown, as a semantic component can be of any type defined in the DynSem
+specification.
+
+As such, it is hard to perform operations such as printing on the returned
+data. One could simply call the `toString` method, but often more control is
```
Comment (Hjdskes) [inline final/discussion/dynsem-data-repr.tex]: \texttt{``toString''}

File: final/discussion/dynsem-data-repr.tex [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+needed, for example if one wants to provide a graphical view of the environment
```
Comment (Hjdskes) [inline final/discussion/dynsem-data-repr.tex]: > needed. For example, ....

File: final/discussion/dynsem-data-repr.tex [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+in the form of a tree structure.
```
Comment (Hjdskes) [inline final/discussion/dynsem-data-repr.tex]: > structure, this solution is inadequate.

**Pull request merged**